File: tools/tests/test_fcs.py
from serial_link.framing import fcs


def test_table_has_256_entries_in_range():
    assert len(fcs.FCSTAB) == 256
    assert all(0 <= v <= 0xFFFF for v in fcs.FCSTAB)


def test_good_fcs_residual():
    # Running the FCS over payload + appended FCS yields the magic constant.
    for payload in (b"", b"\x00", b"\xff\x03\x01\x02\x03", bytes(range(64))):
        framed = payload + fcs.frame_fcs(payload)
        assert fcs.fcs16(framed) == fcs.GOOD_FCS16
        assert fcs.check_fcs(framed)


def test_check_fcs_rejects_corruption():
    payload = b"\xff\x03hello"
    framed = bytearray(payload + fcs.frame_fcs(payload))
    framed[3] ^= 0x01  # flip a bit
    assert not fcs.check_fcs(bytes(framed))


def test_frame_fcs_shape_and_determinism():
    out = fcs.frame_fcs(b"\xff\x03\x21")
    assert isinstance(out, bytes) and len(out) == 2
    assert out == fcs.frame_fcs(b"\xff\x03\x21")  # deterministic

File: tests/test_fcs.cpp
// Mirrors tests/test_fcs.py — FCS-16 table, good residual, corruption rejection.
#include <gtest/gtest.h>

#include "serial_link/fcs.hpp"

using namespace serial_link;
using serial_link::fcs::check_fcs;
using serial_link::fcs::fcs16;
using serial_link::fcs::frame_fcs;

TEST(Fcs, TableHas256EntriesInRange) {
    const auto& tab = fcs::table();
    EXPECT_EQ(tab.size(), 256u);
    for (auto v : tab) EXPECT_LE(v, 0xFFFF);
}

TEST(Fcs, GoodFcsResidual) {
    std::vector<Bytes> payloads = {
        {}, {0x00}, {0xFF, 0x03, 0x01, 0x02, 0x03}, {}};
    Bytes seq64;
    for (int i = 0; i < 64; ++i) seq64.push_back(static_cast<std::uint8_t>(i));
    payloads.back() = seq64;

    for (const auto& payload : payloads) {
        Bytes framed = payload;
        Bytes f = frame_fcs(payload);
        framed.insert(framed.end(), f.begin(), f.end());
        EXPECT_EQ(fcs16(framed), fcs::GOOD_FCS16);
        EXPECT_TRUE(check_fcs(framed));
    }
}

TEST(Fcs, CheckFcsRejectsCorruption) {
    Bytes payload = {0xFF, 0x03, 'h', 'e', 'l', 'l', 'o'};
    Bytes framed = payload;
    Bytes f = frame_fcs(payload);
    framed.insert(framed.end(), f.begin(), f.end());
    framed[3] ^= 0x01;  // flip a bit
    EXPECT_FALSE(check_fcs(framed));
}

TEST(Fcs, FrameFcsShapeAndDeterminism) {
    Bytes out = frame_fcs({0xFF, 0x03, 0x21});
    EXPECT_EQ(out.size(), 2u);
    EXPECT_EQ(out, frame_fcs(Bytes{0xFF, 0x03, 0x21}));
}
